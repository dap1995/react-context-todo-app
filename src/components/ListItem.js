// @flow
import React from 'react';
import styled, { css } from 'styled-components';
import AcceptIcon from '../accept-icon.png';

const Text = styled.span`
  display: flex;
  flex: 8;
  flex-wrap: nowrap;
  padding: 10px;
  transition: color 0.3s linear 0.2s;
  font-size: 24px;
  overflow: scroll;

  ${({ checked }) =>
    checked &&
    css`
      text-decoration: line-through;
      color: #ddd;
      transition: all 0.3s linear;
    `};
`;

const Item = styled.li`
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  overflow: hidden;
  flex-direction: row;
  list-style: none;
  border-bottom: 1px solid #ddd;
  align-items: center;

  &:last-child {
    border-bottom: 0;
  }
`;

const CheckBox = styled.button`
  display: flex;
  flex: 1;
  background-color: white;
  margin: 0.5em;
  min-width: 40px;
  min-height: 40px;
  max-width: 40px;
  max-height: 40px;
  border-radius: 100%;
  background-size: 0%;
  background-image: url(${AcceptIcon});
  background-repeat: no-repeat;
  background-position: center;
  transition: background-size 0.4s cubic-bezier(0.06, 1.74, 0.65, -1.14),
    background-color 0.4s ease 0.2s;

  ${({ checked }) =>
    checked &&
    css`
      background-color: rgba(175, 47, 47, 0.5);
      background-size: 50%;
      transition: background-size 0.4s cubic-bezier(0.6, -1, 0.7, 2),
        background-color 0.4s ease;
    `};
`;

const RemoveButton = styled.button`
  display: flex;
  flex: 1;
  justify-content: center;
  min-width: 10px;
  max-width: 10px;
  margin: 0.5em;
  background-color: transparent;
  border: 0;
  font-size: 24px;
  font-weight: 300;
  color: transparent;
  transition: all 0.2s linear;

  ${Item}:hover & {
    color: #cc9a9a;
    transition: all 0.2s linear;
  }

  ${Item}:hover &:hover {
    color: #af5b5e;
  }
`;

type Props = {
  text: string,
  id: string,
  onClose: (index: string) => void,
  onRemove: (index: string) => void,
  checked: boolean
};

/**
 * Represents an Item Component from a List Component.
 */
class ListItem extends React.PureComponent<Props> {
  onClose = () => {
    const { id, onClose } = this.props;
    onClose(id);
  };

  onRemove = () => {
    const { id, onRemove } = this.props;
    onRemove(id);
  };

  render() {
    const { text, checked } = this.props;
    return (
      <Item>
        <CheckBox onClick={this.onClose} checked={checked} />
        <Text checked={checked}>{text}</Text>
        <RemoveButton onClick={this.onRemove}>{'X'}</RemoveButton>
      </Item>
    );
  }
}

export default ListItem;
