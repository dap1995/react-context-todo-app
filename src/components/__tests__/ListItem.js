import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import ListItem, {
  CheckBox,
  Text,
  RemoveButton,
  InputText,
  Item
} from '../ListItem';

describe('ListItem Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <ListItem
        id={'some-id'}
        text={'some-text'}
        onRemove={() => {}}
        onClose={() => {}}
        checked={false}
      />,
      div
    );
  });

  describe('should calls the function', () => {
    it('onClose when clicking on the checkbox', () => {
      const handleClose = jest.fn();
      const handleRemove = jest.fn();
      const handleEdit = jest.fn();
      const wrapper = mount(
        <ListItem
          id={'some-id'}
          text={'some text'}
          onClose={handleClose}
          onRemove={handleRemove}
          onEdit={handleEdit}
          checked={false}
        />
      );
      const checkbox = wrapper.find(CheckBox);
      checkbox.simulate('click');
      expect(handleClose.mock.calls.length).toBe(1);
      expect(handleRemove.mock.calls.length).toBe(0);
      expect(handleEdit.mock.calls.length).toBe(0);
    });

    it('onRemove when clicking on the removeButton', () => {
      const handleClose = jest.fn();
      const handleRemove = jest.fn();
      const handleEdit = jest.fn();
      const wrapper = mount(
        <ListItem
          id={'some-id'}
          text={'some text'}
          onClose={handleClose}
          onRemove={handleRemove}
          checked={false}
        />
      );
      const removeButton = wrapper.find(RemoveButton);
      removeButton.simulate('click');
      expect(handleClose.mock.calls.length).toBe(0);
      expect(handleRemove.mock.calls.length).toBe(1);
      expect(handleEdit.mock.calls.length).toBe(0);
    });

    it('onEdit when press enter on InputText', () => {
      const handleClose = jest.fn();
      const handleRemove = jest.fn();
      const handleEdit = jest.fn();
      const wrapper = mount(
        <ListItem
          id={'some-id'}
          text={'some text'}
          onClose={handleClose}
          onRemove={handleRemove}
          onEdit={handleEdit}
          checked={false}
        />
      );
      const text = wrapper.find(Text);
      expect(text).toExist();
      text.simulate('dblclick');

      const input = wrapper.find(InputText);
      expect(input).toExist();
      expect(handleEdit.mock.calls.length).toBe(0);

      input.simulate('change', { target: { value: 'some new text' } });
      expect(wrapper.state('text')).toBe('some new text');
      expect(handleEdit.mock.calls.length).toBe(0);

      input.simulate('keydown', { key: 'Enter' });
      expect(handleEdit.mock.calls.length).toBe(1);
      expect(handleClose.mock.calls.length).toBe(0);
      expect(handleRemove.mock.calls.length).toBe(0);
    });

    it('onEdit when blur on InputText', () => {
      const handleClose = jest.fn();
      const handleRemove = jest.fn();
      const handleEdit = jest.fn();
      const wrapper = mount(
        <ListItem
          id={'some-id'}
          text={'some text'}
          onClose={handleClose}
          onRemove={handleRemove}
          onEdit={handleEdit}
          checked={false}
        />
      );
      const text = wrapper.find(Text);
      expect(text).toExist();
      text.simulate('dblclick');

      const input = wrapper.find(InputText);
      expect(input).toExist();
      expect(handleEdit.mock.calls.length).toBe(0);

      input.simulate('change', { target: { value: 'some new text' } });
      expect(wrapper.state('text')).toBe('some new text');
      expect(handleEdit.mock.calls.length).toBe(0);

      const item = wrapper.find(Item);
      item.simulate('blur');
      expect(handleEdit.mock.calls.length).toBe(1);
      expect(handleClose.mock.calls.length).toBe(0);
      expect(handleRemove.mock.calls.length).toBe(0);
    });
  });

  describe('should checkbox and text be', () => {
    it('unchecked', () => {
      const handleClose = jest.fn();
      const handleRemove = jest.fn();
      const wrapper = mount(
        <ListItem
          id={'some-id'}
          text={'some text'}
          onClose={handleClose}
          onRemove={handleRemove}
          checked={false}
        />
      );
      const checkbox = wrapper.find(CheckBox);
      const text = wrapper.find(Text);

      const isCheckBoxChecked = checkbox.prop('checked');
      const isTextChecked = text.prop('checked');

      expect(isCheckBoxChecked).toBeFalsy();
      expect(isTextChecked).toBeFalsy();
      expect(isCheckBoxChecked).toBe(isTextChecked);
    });

    it('unchecked and empty', () => {
      const tree = renderer
        .create(
          <ListItem
            id={'some-id'}
            text={''}
            onClose={() => {}}
            onRemove={() => {}}
            checked={false}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('unchecked and not empty', () => {
      const tree = renderer
        .create(
          <ListItem
            id={'some-id'}
            text={'some text'}
            onClose={() => {}}
            onRemove={() => {}}
            checked={false}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('checked', () => {
      const handleClose = jest.fn();
      const handleRemove = jest.fn();
      const wrapper = mount(
        <ListItem
          id={'some-id'}
          text={'some text'}
          onClose={handleClose}
          onRemove={handleRemove}
          checked={true}
        />
      );
      const checkbox = wrapper.find(CheckBox);
      const text = wrapper.find(Text);

      const isCheckBoxChecked = checkbox.prop('checked');
      const isTextChecked = text.prop('checked');

      expect(isCheckBoxChecked).toBeTruthy();
      expect(isTextChecked).toBeTruthy();
      expect(isCheckBoxChecked).toBe(isTextChecked);
    });

    it('checked and empty', () => {
      const tree = renderer
        .create(
          <ListItem
            id={'some-id'}
            text={''}
            onClose={() => {}}
            onRemove={() => {}}
            checked={true}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('checked and not empty', () => {
      const tree = renderer
        .create(
          <ListItem
            id={'some-id'}
            text={'some text'}
            onClose={() => {}}
            onRemove={() => {}}
            checked={true}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
