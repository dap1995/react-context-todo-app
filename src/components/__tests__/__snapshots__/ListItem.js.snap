// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Component should checkbox and text be checked and empty 1`] = `
<li
  className="sc-ifAKCX jsMcJN"
>
  <button
    checked={true}
    className="sc-EHOje cYIXJ"
    onClick={[Function]}
  />
  <span
    checked={true}
    className="sc-bxivhb ebQrUA"
  >
    
  </span>
  <button
    className="sc-bZQynM gazlcN"
    onClick={[Function]}
  >
    X
  </button>
</li>
`;

exports[`ListItem Component should checkbox and text be checked and not empty 1`] = `
<li
  className="sc-ifAKCX jsMcJN"
>
  <button
    checked={true}
    className="sc-EHOje cYIXJ"
    onClick={[Function]}
  />
  <span
    checked={true}
    className="sc-bxivhb ebQrUA"
  >
    some text
  </span>
  <button
    className="sc-bZQynM gazlcN"
    onClick={[Function]}
  >
    X
  </button>
</li>
`;

exports[`ListItem Component should checkbox and text be unchecked and empty 1`] = `
<li
  className="sc-ifAKCX jsMcJN"
>
  <button
    checked={false}
    className="sc-EHOje fdMPqF"
    onClick={[Function]}
  />
  <span
    checked={false}
    className="sc-bxivhb btjxdt"
  >
    
  </span>
  <button
    className="sc-bZQynM gazlcN"
    onClick={[Function]}
  >
    X
  </button>
</li>
`;

exports[`ListItem Component should checkbox and text be unchecked and not empty 1`] = `
<li
  className="sc-ifAKCX jsMcJN"
>
  <button
    checked={false}
    className="sc-EHOje fdMPqF"
    onClick={[Function]}
  />
  <span
    checked={false}
    className="sc-bxivhb btjxdt"
  >
    some text
  </span>
  <button
    className="sc-bZQynM gazlcN"
    onClick={[Function]}
  >
    X
  </button>
</li>
`;
